### ISSUE TITLE
mapToObj is not type safe

### ISSUE THREAD

**Issue by jasonkuhrt** [remeda/remeda#384]
Here is a bug reproduction:

```ts
import { mapToObj } from 'remeda'

interface FeatureFlagA {
  name: 'a'
  rule: 'allow' | 'deny'
}
interface FeatureFlagB {
  name: 'b'
  rule: 'allow' | 'deny'
}
interface FeatureFlagC {
  name: 'c'
  rule: 'allow' | 'deny'
}
interface User {
  featureFlags: (FeatureFlagA | FeatureFlagB | FeatureFlagC)[]
}
const user: User = {
  featureFlags: [
    { name: `a`, rule: `allow` },
    { name: `b`, rule: `deny` },
  ],
}

const featureFlags = mapToObj(user.featureFlags, (_) => [_.name, _.rule === `allow`])

// Wrong!
featureFlags.c // typeof = boolean

console.log(Object.entries(featureFlags))
```
```
❯ pnpm tsx flaw.ts
[ [ 'a', true ], [ 'b', false ] ]
```

The correct type for `featureFlags.a/b/c` should be `a/b/c?: boolean`.

This is because it is non-deterministic whether or not `user.featureFlags` will have an instance of every possible type.

**Comment by TkDodo** [comment]
I think the problem is that you are widening your `user` object to be of type `User`. To get a better, more narrow type inference, you would need to use the `satisifes` operator:

```
const user = {
  featureFlags: [
    { name: `a`, rule: `allow` },
    { name: `b`, rule: `deny` },
  ],
} satisfies User
```

see [this playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbziAhmAKhA8gIwFZwC+cAZlBCHAORQCmItAJilQFCvAB2MtUJKAY1pwAYrRQwArnREAbFAHMAgolZw4nFAwBc1FmrhRJs2rqopZsiAHcqcAD7VGtTgE82hDt178ho8VIy8goAQqrqmjrUOGzqRiZmFla2Dk4u7qyeXDx8gsJiEtK0cooAwuEaWqbUArGGxtXmljZ2jlTObh5eOb7CAKoAzrwVJAFFJQoDugAUBYHFwSqOc+PBYctjQWUAlADaALqZrAIQnAPwkkNQcAC8I5sLilNwuwbqSJHVAAYoXwA09QScB+zWsXyIfzeiEqUS+OH+gO+HVc4MIkPU+0hxAGEmAAxIwFoAzgg147BOZ3go0KW0mt2QaEwuDw00uvAAdNT5hMBgDpgB9ba3AB8L357M+APF8WENzlwKSNi++227AA9Gq4AB1cicBQAQlYXNWT3ZAjgGrgMFcYFoEBI9JwEAgJhQnHJpwGLto7KsCmmzNoAhg7JcMCghIG02NtIG222QA)

**Comment by TkDodo** [comment]
and you will get the same problem using the built-in `Object.fromEntries`:

```
const featureFlags = Object.fromEntries(user.featureFlags.map((_) => [_.name, _.rule === `allow`]))

// Wrong!
featureFlags.c // typeof = boolean
```

except that this is very widely typed to `Record<string, boolean>`, so you could do `featureFlags.foobar` as well and would get `boolean` back

**Comment by jasonkuhrt** [comment]
> I think the problem is that you are widening your user object to be of type User.

I get your point, but no, it was intentional. Imagine for example this data is coming from the database.

> and you will get the same problem using the built-in Object.fromEntries:

Sure, hence we reach for tools like `remeda` right? : )

**Comment by TkDodo** [comment]
Yeah, maybe what we could do is infer them as `Partial`. I guess there's a reason why `Object.fromEntries` infers an index signature. At least with `noUncheckedIndexedAccess`, you would get `boolean | undefined` for everything.

Thinking about it, I guess the current behaviour should go into `mapToObj.strict` and the normal version could always return a Partial, but not sure. @eranhirsch what do you think ?

**Comment by jasonkuhrt** [comment]
I would expect `.strict` to produce the maximally type-safe results.

What I'm showing above is simply a bug. Logically, a list of objects is non-deterministic and it is impossible to guarantee a derived enum structure from it at runtime.

We built this utility function in our codebase for the use-case:

```ts
const featureFlags = mapValues(
    indexByEnum(
      prismaUser.experimentalFeatureAccessPolicies,
      `feature`,
      keysOfStrict(PrismaTypes.$Enums.ExperimentalFeature),
    ),
    (_) => _?.rule === `allow`,
  )
```

1. Accept the non-deterministic structure
2. Accept the field name to build the index from
3. Accept the field _values_ which constitute the _enum member space_

The implementation of `indexByEnum`

```ts
/**
 * Index an array of objects by an enumerated field shared by all elements.
 *
 * The index is guaranteed to contain a key for each enumeration, however the value is possibly
 * undefined since array is a non-deterministic data structure.
 */
export const indexByEnum = <T extends object, FN extends keyof T, FV extends TuplifyUnion<T[FN]>>(
  values: T[],
  fieldName: FN,
  fieldValues: FV,
): { [k in FV[number] & string]?: T } => {
  type IndexFN = FV[number] & string
  const index: { [k in IndexFN]?: T } = {}
  for (const fieldValue of fieldValues as IndexFN[]) {
    index[fieldValue] = values.find((_) => _[fieldName] === fieldValue)
  }
  return index
}
```

**Comment by eranhirsch** [comment]
This is because of how typescript differentiates infinite records and finite ones. With the strictest typescript settings, an infinite record is implicitly `Partial`, but with looser typescript settings it's actually `Required`. Whereas for finite records (with literal union keys), the definition is more consistent for both strict and unstrict typescript.

https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&allowUnreachableCode=true&allowUnusedLabels=true&noUnusedLocals=true&noUnusedParameters=true&target=99&jsx=1&useUnknownInCatchVariables=true&exactOptionalPropertyTypes=true&noFallthroughCasesInSwitch=true&noImplicitOverride=true&noPropertyAccessFromIndexSignature=true#code/MYewdgzgLgBAlmAZguUCmAuGAlNoBOAJgDwAGcAtgIboUAOANjWgCQDe0+CA5gL6kAaGACMQIBmipgAfDAC8MNrwDcAKFCRYyMKkw48IIsQBEwGsZgAfGMcIhuxoaPGSZ8xTDNQsUfAFc0ITtuLEQqBgg0GBVVdXBoGAAPdwRtXQBtAHJKZnomdAALNAYGEAB3QwZCTIBdNQB6ephmmAA9AH5YjQSAT3c09AA6Lwamlo7YuM0YOip8KDhwgEkkFHQsAAU5hfDiKB66NBBEeFWddFkFJTVu2Fn5xYYAMTW9LYfd-cPjmAG0S8UMSmCUSACZ3PcdgwVn8sjlaIxmEUSuVKtU6qpGi02p1gbAeuCFJDHi9zmhhjRRtiOkA

This means that the function should have different return signatures based on whether the generic `K` is finite or not.

For the broadest definition (e.g. if `K` is just `string`) we can detect this by simply checking that `string extends K`, but for string templates (e.g. if `K` is `hello${number]`) it won't be that easy and we would return a type that might be wrong for some people...

I'm not sure what the best solution would be.

**Comment by russelldavis** [comment]
@eranhirsch consistent with your analysis, I don't see a way to fix this without a breaking change. Looks like a new major version is in beta, so that could be a good place to introduce it.

I think @TkDodo's idea of making it `Partial` is the way to go. Yes, it essentially forces `noUncheckedIndexedAccess` for this one function, but that's also the safer and more correct behavior anyway.

**Comment by russelldavis** [comment]
Looking at this closer, your idea of treating infinite vs finite records differently is strictly better, since it only creates a breaking change for `noUncheckedIndexedAccess: false` users for keys types that you can't properly distinguish, like template literal types. So +1 to that.

It looks like you implemented this idea in the beta branch with the `ExactRecord` type, and are using it in places like `mapKeys`, so we'd just need to use that type on `mapToObj` as well.

**Comment by eranhirsch** [comment]
We are considering deprecating mapToObj (because it's kinda like `fromEntries` anyway, and we now also have `fromKeys` and `pullObject`).

If we don't we'll probably improve it's return type in v2.

**Comment by eranhirsch** [comment]
@cjquines - What do you think about dropping this utility in v3? It's equivalent to `fromEntries(map(DATA, mapper))`, and fixing it would mean duplicating and repeating a lot of those utilities' typing magic without gaining anything unique to this function.

Am I missing something in terms of runtime utility, or in what kind of types we can infer over the composition that makes this worthwhile keeping?

**Comment by eranhirsch** [comment]
The very minimal (runtime) testing we have for this utility passes when changing the implementation to the composition of fromEntries and map, it actually even makes more sense when inside a pipe:

```ts
const result = pipe(
  [1, 2, 3],
  mapToObj((x) => [String(x), x * 2]),
);

// Becomes:
const result = pipe(
  [1, 2, 3],
  map((x) => [Stirng(x), x * 2] as const),
  fromEntries(),
);
```

**Comment by eranhirsch** [comment]
Actually there is a slight performance benefit by using this function over the composition of `map` and `fromEntries` considering that `fromEntries` isn't lazy, which would require us to generate the whole array of tuples via `map` before they could be converted to an object; this is in contrary to `mapToObj` which would create the object lazily without the intermediate array. `pullObject` is a way around this, but requires that the logic for keys and the logic for values don't repeat each other, and does add an additional function invocation per item in the input array.

These are all extreme edge cases, but they are non-trivial in those instances. I'll modify the PR so that the function isn't deprecated per se in v2, but that once we implement `fromEntries` to work lazily in v3, it'll be deprecated and removed.
